xmtp_mls/src/utils/cleanup_duplicate_updates.rs

```diff
@@ -0,0 +1,231 @@
+    group_message::{ContentType, MsgQueryArgs},
+    prelude::QueryGroupMessage,
+};
+
+use crate::groups::mls_sync::GroupMessageProcessingError;
+use crate::messages::decoded_message::MessageBody;
+use crate::messages::enrichment::enrich_messages;
+
+const BATCH_SIZE: i64 = 100;
+
+pub async fn perform<C>(db: DbConnection<C>)
+where
+    C: ConnectionExt,
+{
+    if let Err(err) = perform_inner(db).await {
+        tracing::error!("Duplicate cleanup task failed: {err:?}");
+    }
+}
+
+async fn perform_inner<C>(db: DbConnection<C>) -> Result<(), GroupMessageProcessingError>
+where
+    C: ConnectionExt,
+{
+    let prefs = StoredUserPreferences::load(&db)?;
+    if prefs.dm_group_updates_migrated {
+        info!("DM group updates migration has already been performed. Skipping.");
+        return Ok(());
+    }
+
+    let mut group_offset = 0;
+    let mut groups;
+    loop {
+        groups = db.find_groups_by_id_paged(
+            GroupQueryArgs {
+                conversation_type: Some(ConversationType::Dm),
+                limit: Some(BATCH_SIZE),
+                ..Default::default()
+            },
+            group_offset,
+        )?;
+
+        if groups.is_empty() {
+            break;
+        }
+
+        for group in groups {
+            let mut sent_after_ns = None;
+            let mut msgs;
+            let mut originals: HashSet<u64> = HashSet::default();
+
+            loop {
+                msgs = db.get_group_messages(
+                    &group.id,
+                    &MsgQueryArgs {
+                        content_types: Some(vec![ContentType::GroupUpdated]),
+                        sent_after_ns,
+                        limit: Some(BATCH_SIZE),
+                        ..Default::default()
+                    },
+                )?;
+
+                {
+                    let Some(msg) = msgs.last() else {
+                        break;
+                    };
+                    sent_after_ns = Some(msg.sent_at_ns);
+                }
+
+                let msgs = enrich_messages(&db, &group.id, msgs)?;
+
+                for msg in msgs {
+                    let MessageBody::GroupUpdated(update) = msg.content else {
+                        continue;
+                    };
+
+                    let mut hasher = DefaultHasher::new();
+                    update.added_inboxes.hash(&mut hasher);
+                    update.removed_inboxes.hash(&mut hasher);
+                    update.metadata_field_changes.hash(&mut hasher);
+                    if originals.insert(hasher.finish()) {
+                        continue;
+                    }
+
+                    db.raw_query_write(|conn| {
+                        xmtp_db::diesel::delete(xmtp_db::schema::group_messages::table)
+                            .filter(xmtp_db::schema::group_messages::id.eq(&msg.metadata.id))
+                            .execute(conn)
+                    })?;
+
+                    tokio::task::yield_now().await;
```

Comment:
Manually yield, since we don't have any awaits in this potentially long running task and it could block wasm until it's finished.